<p style="align: center;"><img src="https://static.tildacdn.com/tild6636-3531-4239-b465-376364646465/Deep_Learning_School.png", width=500></p>

<h3 style="text-align: center;"><b>Физтех-Школа Прикладной математики и информатики (ФПМИ) МФТИ</b></h3>

---

<h2 style="text-align: center;"><b>Object detection</b></h2>

### Руководители проекта: 
* Илья Захаркин (Samsung AI Center - Moscow) | tg:@izakharkin
* Юрий Яровиков (AIRI, МФТИ) | tg:@yu_rovikov

<h2 style="text-align: center;"><b>Этапы работы</b></h2>

Проект подразумевает выполнение всеми участниками пунктов 1-2, и далее выбор каждым слушателем своего сценария -- 1 или 2. Можно сделать оба сценария, то есть обучить модель на своих данных и её же встроить в демо, но для этого стоит быть готовым к удвоению затрачиваемого на проект времени.

<h3 style="text-align: center;"><b>Общий этап работы</b></h3>

**Hint**: если у вас на компьютере нет GPU, то лучше с самого начала делать все в Google Colab.

**1). Выбор фреймворка/библиотеки для использования детектора**

Чтобы освежить память о задаче детекции, можно посмотреть [занятия на продвинутом курсе](https://stepik.org/lesson/458312/step/1?unit=616130).

В выборе фреймворка предоставляется свобода, лично я рекомендовал бы один из:
- `torchvision.models.detection` и `torchhub`: "нативные" модели для детектирования прямо из PyTorch. Примеры использования есть прямо на занятиях DLSchool по практике CV [2019 года](https://www.youtube.com/watch?v=XSPYe4-y4HE) и [2020 года](https://stepik.org/lesson/458313/step/1?unit=616131);
- `mmdetection`: как с ним работать, рассказывается в [практическом занятии](https://stepik.org/lesson/458313/step/2?unit=616131).
- `detectron2`: краткая информация есть в конце [занятия DLSchool по практике CV](https://www.youtube.com/watch?v=XSPYe4-y4HE), можно начать с него. Лучше самостоятелньо изучить [официальный репозиторий](https://github.com/facebookresearch/detectron2) и уже с ним работать в дальнейшем ("Quick Start");
- `TensorFlow Object Detection API`: как с ним работать рассказывается в [занятии 2018 года](https://www.youtube.com/watch?v=xHIzyrU1uVM). Работать предстоит с [официальным репозиторием](https://github.com/tensorflow/models/tree/master/research/object_detection).

Рекомендую выбрать один из них и работать уже с этим модулем.

Для работы с детектором выбрана библиотека `ultralytics`, которая предоставляет инструменты для использования семейства моделей *YOLO*, в частности новейшее *YOLOv8* поколение. Кроме того, для работы с веб-камерой используются средства библиотеки `OpenCV`.

In [2]:
# !pip install ultralytics

**2). Запуск детектора на случайных изображениях**

Этот пункт просто про то, чтобы запустить любую модель детектирования в выбранном выше репозитории. Таким образом, часть с запуском будет работать, и далее уже можно приступать к основным сценариям.

Поскольку модель обучена на наборе данных *COCO128* подберем картинки с подходящими классами. Для детекции используем *YOLOv8 medium* модель.

__P.S. Нотбук запускается из рабочей среды проекта, поэтому установка зависимостей не описывается.__

In [4]:
!yolo detect predict model=yolov8m.pt source=img

Ultralytics YOLOv8.0.18 🚀 Python-3.10.6 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce GTX 1650, 3904MiB)
YOLOv8m summary (fused): 218 layers, 25886080 parameters, 0 gradients, 78.9 GFLOPs
image 1/4 /home/danil/Projects/detector-demo/img/animals.jpg: 384x640 1 bird, 4 cats, 5 dogs, 30.6ms
image 2/4 /home/danil/Projects/detector-demo/img/city.jpg: 448x640 7 persons, 2 bicycles, 12 cars, 1 bus, 2 traffic lights, 1 fire hydrant, 1 backpack, 1 umbrella, 1 handbag, 34.4ms
image 3/4 /home/danil/Projects/detector-demo/img/pc.jpg: 448x640 1 chair, 2 tvs, 1 mouse, 1 keyboard, 32.8ms
image 4/4 /home/danil/Projects/detector-demo/img/room.jpg: 480x640 1 bowl, 2 chairs, 3 couchs, 1 bed, 8 books, 35.6ms
Speed: 0.5ms pre-process, 33.3ms inference, 2.1ms postprocess per image at shape (1, 3, 640, 640)
Results saved to runs/detect/predict


|![alt](runs/detect/predict/pc.jpg) |![alt](runs/detect/predict/animals.jpg)|
|-|-|
|![alt](runs/detect/predict/room.jpg) | ![alt](runs/detect/predict/city.jpg)

<h3 style="text-align: center;"><b>1 сценарий</b></h3>

В первом сценарии упор делается на разработку демо, в которое встривается нейросетевая модель для детекции.

Если у Вас есть опыт веб- или мобильной разработки, то можете использовать именного его и работать в рамках привычных Вам инструментов. Главное, чтобы в итоге они позволяил встроить в себя нейросетевой детектор, на вход которому будут поступать картинки.

Изображения на вход демо могут поступать с веб-камеры, из файлов, по ссылке или с камеры мобильного телефона -- способ может быть любой (на Ваш выбор). Демо должно показывать, что детектор успешно отрабатывает на поданных изображениях и находит нужные объекты. Оформлять красиво визуальную часть демо -- приветствуется, но не обязательно -- это может быть просто кнопка "загрузить" и показ фото до/после детекции.

**3). Выбор фреймворка/библиотеки для разработки веб/мобильного демо**

Основным инструментом для разработки веб-демо будет микрофреймворк **Flask**: [серия туториалов на русском](https://habr.com/ru/post/346306/).   
Полезные ресуры:
- [курс по веб-разработке](https://www.youtube.com/playlist?list=PLzQrZe3EemP5KsgWGnmC0QrOzQqjg3Kd5), нас интересуют первые 7 видео в плейлисте. В частности, нужны видео по Flask, там очень хорошие обучалки параллельно с лектором;
- [исчерпывающий справочник по Flask (англ)](https://www3.ntu.edu.sg/home/ehchua/programming/webprogramming/Python3_Flask.html);
- можно посмотреть мой [репозиторий с реализацией веб-демо](https://github.com/izaharkin/Respalyzer) для ML-задачи оценки отзывов.

Для разработки мобильного демо стоит выбрать инстурмент на свое усмотрение:
- под Android: [пример на Pytorch Mobile](https://towardsdatascience.com/object-detector-android-app-using-pytorch-mobile-neural-network-407c419b56cd), [пример на TensorFlow Lite](https://www.tensorflow.org/lite/models/object_detection/overview). **Примечание** от Дмитрия Шумилина: на Android с TF Lite на момент января 2021 есть [ошибка](https://github.com/tensorflow/models/issues/9341) с новым форматом хранения модели. Можно попробовать возможное [решение](https://www.youtube.com/watch?v=syTKGY-H44E&ab_channel=DoomsdayRobotics) или писать на чистом Java. Также можно попробовать использовать более старые версии TensorFlow, в которых проблем совместимости еще не было, например, [v2.1.0](https://github.com/tensorflow/tensorflow/releases/tag/v2.1.0).
- под iOS: [пример на TensorFlow Lite Swift API](https://www.tensorflow.org/lite/guide/inference#load_and_run_a_model_in_swift), [пример с Vision Framework](https://developer.apple.com/documentation/vision/recognizing_objects_in_live_capture) на "чистом" Swift'е.

Разумеется, лучше **самостоятельно поискать видео/статьи** на тему использования моделей на мобильных устройствах.

Для разработки веб демо выбран фреймворк `streamlit`, предоставляющий готовые решения "из коробки" для создания и развертывания собственного веб сервиса.

**4). Разработка демо**

Этот пункт про сам процесс написания кода для демо. 

> Результатом пункта является код, который можно запустить. Не хватать будет только логики детектора, сам интерфейс должен быть уже рабочим.

Веб интерфейс реализован в файле [app.py](https://github.com/DaaN1L/object-detection-demo/blob/main/app.py) рабочей директории.

**5). Встраивание модели-детектора в демо**

Этот пункт про процесс дописывания кода, который будет обеспечивать "логику" демо $-$ само детектирование.

> Результатом пункта является код, который можно запустить и продемонстрировать работающую систему детектирования объектов.

Логика детектора описана в файле [predictor.py](https://github.com/DaaN1L/object-detection-demo/blob/main/predictor.py) рабочей директории.

**6). Тестирование демо**

Здесь нужно запустить ваше демо на как можно большем количестве примеров, чтобы понять, в чем его сильные и слабые стороны. То есть какие объекты/сцены детектор обрабатывает легко, а с какими ему справится сложно. Нужно предложить также пути для улучшения модели на основе увиденных ошибок.

> Результатом пункта является набор изображений, на которых демо отработало. Для каждого изображения нужно добавить комментарии, почему модель справилась хорошо/плохо, предложить пути ее улучшения.

[Видео с тестированием](https://drive.google.com/file/d/1af6hQraDEKzsl6dy0TwMaI_cox_pWxo9/view?usp=sharing)

<img src=https://raw.githubusercontent.com/ultralytics/assets/main/yolov8/yolo-comparison-plots.png>

<center><b>Detection Summary</b></center>

<table>
<thead>
<tr>
<th>Model</th>
<th>size<br><sup>(pixels)</sup></th>
<th>mAP<sup>val<br>50-95</sup></th>
<th>Speed<br><sup>CPU ONNX<br>(ms)</sup></th>
<th>Speed<br><sup>A100 TensorRT<br>(ms)</sup></th>
<th>params<br><sup>(M)</sup></th>
<th>FLOPs<br><sup>(B)</sup></th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8n.pt">YOLOv8n</a></td>
<td>640</td>
<td>37.3</td>
<td>80.4</td>
<td>0.99</td>
<td>3.2</td>
<td>8.7</td>
</tr>
<tr>
<td><a href="https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8s.pt">YOLOv8s</a></td>
<td>640</td>
<td>44.9</td>
<td>128.4</td>
<td>1.20</td>
<td>11.2</td>
<td>28.6</td>
</tr>
<tr>
<td><a href="https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8m.pt">YOLOv8m</a></td>
<td>640</td>
<td>50.2</td>
<td>234.7</td>
<td>1.83</td>
<td>25.9</td>
<td>78.9</td>
</tr>
<tr>
<td><a href="https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8l.pt">YOLOv8l</a></td>
<td>640</td>
<td>52.9</td>
<td>375.2</td>
<td>2.39</td>
<td>43.7</td>
<td>165.2</td>
</tr>
<tr>
<td><a href="https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8x.pt">YOLOv8x</a></td>
<td>640</td>
<td>53.9</td>
<td>479.1</td>
<td>3.53</td>
<td>68.2</td>
<td>257.8</td>
</tr>
</tbody>
</table>
<ul dir="auto">
<li><strong>mAP<sup>val</sup></strong> values are for single-model single-scale on <a href="http://cocodataset.org" rel="nofollow">COCO val2017</a> dataset.
<br>Reproduce by <code>yolo val detect data=coco.yaml device=0</code></li>
<li><strong>Speed</strong> averaged over COCO val images using an <a href="https://aws.amazon.com/ec2/instance-types/p4/" rel="nofollow">Amazon EC2 P4d</a>
instance.
<br>Reproduce by <code>yolo val detect data=coco128.yaml batch=1 device=0/cpu</code></li>
</ul>

**7). Оформление демо для показа другим людям**

В этом пункте можно пойти двумя путями:
1. Проделать работу по улучшению визуальной составляющей демо (интерфейс)
2. Загрузить модель на какой-нибудь сервер/хост/test-flight (в случае мобильного iOS-демо), чтобы к демо можно было обратиться прямо в адресной строке браузера

\> По *первому пункту* могу посоветовать использовать библиотеку [Bootstrap](https://habr.com/ru/post/349060/), для мобильного демо элементы UI/UX являются частью основной разработки (поэтому стоит просто погуглить/почитать документацию).

\> *Второй пункт - в случае веб-демо*: 

Способ 1: Google Cloud Engine.

Если ваше приложение требует установки системных пакетов, например, через `apt-get install`, то вам придется работать на выделенном сервере VPN или на виртуальной машине. К счастью тот же [Google Cloud](https://cloud.google.com/compute) предоставляет бесплатные 300$ на 90 дней использования Виртуальной машиной, чего хватит в большинстве случаев. Эти ссылки помогут вам понять, как в таком случае создать виртуальную машину, установить и настроить виртуальное окружение и вебсервер, а также задеплоить проект:

- [Deploying a Flask app to a Virtual Machine](https://www.youtube.com/watch?v=a2g9pDleGQk&ab_channel=JulianNash)
- [Set up Gunicorn and Nginx](https://www.digitalocean.com/community/tutorials/how-to-serve-flask-applications-with-gunicorn-and-nginx-on-ubuntu-20-04-ru)

Способ 2: Heroku.

Если с GCE проблемы/не хочется привязывать карту и т.д., могут помочь эти ресурсы и сервис [Heroku](https://www.heroku.com/):
- [Flask deployment](http://www.tutorialspoint.com/flask/flask_deployment.htm)
- [Deploy Flask app to Heroku (youtube)](https://www.youtube.com/watch?v=pmRT8QQLIqk)
- [Deploy Flask app to Heroku (medium)](https://medium.com/the-andela-way/deploying-your-flask-application-to-heroku-c99050bce8f9)
- [Set your own domain name on Heroku](https://devcenter.heroku.com/articles/custom-domains)

\> *Второй пункт - в случае мобильного демо*: 

Здесь с как таковым деплоем сложнее, обычно мобильные приложения публикуются или в Google Play (Android), или в AppStore (iOS). Однако можно снять **видеопоказ экрана (скринкаст)** с использованием написанного приложения - вполне подойдет для публичной демонстрации.

<h3 style="text-align: center;"><b>2 сценарий</b></h3>

Во втором сценарии упор делается на сбор выборки, преобработку данных, обучение модели и измерение качества её работы. То есть осуществляется полный цикл разработки нейросетевой модели для задачи детектирования без встривания в веб/мобильный прототип.

**3). Выбор датасета**

При работе с датасетом вы неизбежно столкнетесь с работой с файлами и папками (директориями). Рекомендуется освежить в памяти работу с библиотеками `os`, `json`, `glob`. Может помочь [этот туториал](https://realpython.com/working-with-files-in-python/).

На выбор предоставляются 5 датасетов по детекции объектов:
1. [Детекция игровых карт](https://github.com/EdjeElectronics/TensorFlow-Object-Detection-API-Tutorial-Train-Multiple-Objects-Windows-10): лежат в папке images вместе с разметкой;
2. [Детекция фруктов](https://www.kaggle.com/mbkinaci/fruit-images-for-object-detection): скачать можно, нажав на кнопку Download;
3. [Детекция одежды (Deep Fashion 2)](https://github.com/switchablenorms/DeepFashion2): стоит прочитать README на главной странице репозитория. Для получения датасета нужно запросить пароль у автора через гугл-форму. После скачивания распакуйте его с использованием пароля. Из файлов аннотаций нас будут интересовать только `bounding_box`, `category_name` и `category_id`;
4. [Детекция лиц (Wider Face)](http://shuoyang1213.me/WIDERFACE/): большой датасет для детектирования лиц самых разных размеров. Скачать можно прямо по ссылкам на сайте;
5. [Детекция лиц (Kaggle)](https://www.kaggle.com/dataturks/face-detection-in-images): в датасете достаточно мало данных, но можно попробовать, если датасеты выше показались неподходящими для Вас;
6. Датасет из любого соревновани по детекции на Kaggle.

> Результатом выполнения пункта явлется загруженный датасет, состоящий из изображений и разметки к ним (bounding box'ов всех объектов на каждом изображении).

**4). Предобработка данных**

Самый непростой этап в этом сценарии. Скачать данные $-$ лишь половина дела. Чтобы обучить нейросеть на этих данных, нужно написать генератор батчей. Однако если будем подавать изображения так, как они есть, то даже батч собрать не сможем -- нужно привести их к однмоу размеру. Далее нужно привести их к типу float, переместить на CUDA и поделить значения в пикселях на 255 (подробнее см. [занятие](https://www.youtube.com/watch?v=XSPYe4-y4HE)). Также нужно настроить аугментации и постобработку.

То, как именно все это реализовать $-$ зависит от инструмента, выбранного в пункте 1. Например, в detectron2 в обучающих материалах описан формат данных для обучения. Возможно, нужно будет зайти в документацию и почитать более подробно, чтобы разобраться, какой именно нужен формат координат.

НЕ нужно копировать все файлы с картинками и разметкой прямо на диске в их предобработанные версии. Хороший тон $-$ осуществлять всю эту обработку программно, "на лету". Поможет [туториал](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html) по написанию своего датасета на PyTorch.

> Результатом выполнения пункта явлется код, запуск которого ведет к подаче батчей правильного вида (разметка приведена к требуемому формату координат, изображения нужного типа, размера и поделены на 255 и т.д.) для обучения нейронной сети-детектора.

**5). Обучение модели-детектора** 

Стоит написать цикл обучения на PyTorch самостоятельно, однако допускается использовать [Catalyst](https://github.com/catalyst-team/catalyst), [PyTorch Lightning](https://github.com/williamFalcon/pytorch-lightning) или [Ignite](https://github.com/pytorch/ignite). Например, саму модель можно взять из фреймворка из пункта 1, а обучение осуществлять с помощью одной перечисленных трех библиотек.

> Результатом выполнения пункта явлется код, запуск которого ведет к обучению модели на выбранном датасете. При обучении **обязательно выводить числовые значения лосса на трейне и валидации**, крайне желательно использовать [`TensorBoard`](https://pytorch.org/docs/stable/tensorboard.html) для визуализации. Обязательно также сохранять модель после каждой N-ой эпохи, чтобы потом ее качество можно было проверить и веса были переиспользуемыми.

**6). Измерение качества работы модели (метрики согласуются с руководителем и зависят от задачи)**

Под метриками понимаются функции/формулы, по которым оценивается качество модели-детектора. Обычно для измерения качества работы детектора используют поклассовые Precision, Recall, F1-меру и mean Average Precision (mAP). Подробнее про них можно послушать в [видеолекции 2018 года](https://www.youtube.com/watch?v=ewkSI2cuyoQ&list=PL0Ks75aof3ThkitsZbUOEQg7Ybl5kB_s3&index=24).

Необходимо самостоятельно реализовать функции, вычисляющие поклассовые Precision и Recall. На вход функциям поступают веса модели и выборка, на которой нужно измерить метрики. На выходе $-$ таблица с precision и recall для каждого класса.

> Результат пункта $-$ реализованные функции метрик для задачи детектирования, позволяющие оценить качество работы модели на выборке. Выборка/даталоадер является аргументом функций.

 **7). Поиск путей применения этой модели в бизнесе/реальных задачах/набросок встраивания в веб/мобильное демо**
 
В этом пункте нужно подумать, как эта модель может быть использована в дальнейшем. То есть, например, зачем нужно детектировать фрукты? Или одежду?

> Результат пункта $-$ перечисленные кейсы использования модели (описанные **как можно подробнее**).

**IMPORTANT NOTE:** Обычно этим вопросом все же задаются до начала какой-либо разработки. Просто в данном случае проект выполняется скорее в образовательных целях.

<h2 style="text-align: center;"><b>Критерии оценивания</b></h2>

* 1 пункт $-$ 1 балл 
* 2 пункт $-$ 1 балл 
* 3 пункт $-$ 0 баллов (промежуточный пункт)  
* 4 пункт $-$ 3 балла   
* 5 пункт $-$ 3 балла   
* 6 пункт $-$ 1 балл 
* 7 пункт $-$ 1 балл 
* Максимум баллов по проекту $-$ 10  

**Успехов в выполнении проекта!** 

Желаю всем проделать полезную, интересную и качественную работу, которую потом нестыдно и в резюме указать, и друзьям показать ;)